scripts/airgapped/get-all-images.py

```diff
@@ -0,0 +1,177 @@
+                         (GET_IMAGES_SH, app["charm"], stderr.decode("utf-8"))
+                         )
+    images = stdout.decode("utf-8").split("\n")
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_dependency_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a dependency charm by:
+    1. Cloning the repo of the charm
+    2. Looking at its metadata.yaml for "upstream-source" keys
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_bundle_images(bundle_dict: dict) -> list[str]:
+    """Return a list of images used by a bundle"""
+    images = []
+
+    for app_name, app in bundle_dict["applications"].items():
+        logging.info(f"Handling app {app_name}")
+
+        # exclude repos we know don't have images
+        if app_name in EXCLUDE_REPOS:
+            logging.info("Ignoring charm %s", app["charm"])
+            continue
+
+        # Follow default image-gather process for dependency apps
+        if is_dependency_app(app):
+            logging.info("Dependency app '%s' with charm '%s'", app_name,
+                         app["charm"])
+            images.extend(get_dependency_app_images(app))
+            continue
+
+        # Ensure analytics app has necessary github repo/branch metadata
+        if not bundle_app_contains_gh_metadata(app):
+            raise KeyError(
+                "Application '%s' doesn't include expected gh metadata keys" %
+                app_name
+            )
+
+        images.extend(get_analytics_app_images(app))
+
+    return images
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Gather all images from a bundle"
+    )
+    parser.add_argument("bundle")
+    parser.add_argument("--airgap-testing", action="store_true")
+
+    args = parser.parse_args()
+
+    bundle_dict = {}
+    with open(args.bundle, 'r') as file:
+        bundle_dict = yaml.safe_load(file)
+
+    # keep unique images and sort them
+    images_set = set(get_bundle_images(bundle_dict))
+    if '' in images_set:
+        images_set.remove('')
+
+    images = list(images_set)
+    images.sort()
+
+    # append the airgap images
+    if args.airgap_testing:
+        images.extend(AIRGAP_TESTING_IMAGES)
+
+    logging.info(f"Found {len(images)} different images")
+
+    for img in images:
+        print(img)
```

Comment:
I'm thinking this is the final list of images that users would be interested in, so instead of just printing them, should we store them in an output dir?

One thing you could do is add a `--output` arg that users can pass to define the output file for all the images.

Reply:
I was thinking about this as well when writing the first iteration of the scripts. In the end I thought that users can just redirect the final output to a file with `python3 script.py > images.txt` and avoided creating logic for this

Reply:
Let's use the comment below for this discussion
https://github.com/canonical/bundle-kubeflow/pull/1038#discussion_r1741887893